### PR TITLE
Project Authentication

### PR DESCRIPTION
**Pull request by OlukaDenis**
- [x] Sign in and signout
- [x] Forms
- [x] Sessions
- [x] Validations
- [x] Remember me

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,3 +1,47 @@
...
+    return false if remember_digest.nil?
+
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  end
+
+  private
+
+  def new_remember_token
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
```
Comment (maelfosso) [inline app/models/user.rb:45]: That line causes me a lot of trouble. I was obliged to comment it to be able to run your project.

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+module SessionsHelper
+  # logs in the user
+  def log_in(user)
+    session[:user_id] = user.id
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb:6]: Don't use `session`

- Generate and Update the user `remember_token` here
- And then call you `remember(user)` method

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required

Hello @OlukaDenis 

You did a really nice job. However, there are a some changes that you need to do to move forward.

1. I was not easy to run your code. Read the comment I left below.

2. Please, forgot about using `session`. Only think about `cookies`.
Screenshot from 2019-12-12 18-19-27
Refer to that screenshot.

Kindly make the changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user&.authenticated?(cookies[:remember_token])
+        log_in user
+        @current_user = user
+      end
+    end
+  end
```
Comment (OlukaDenis) [inline app/helpers/sessions_helper.rb:20]: I really don't understand why I am not supposed to use sessions in this case. 
Sessions will be used temporarily when a user doesn't tick the remember me checkbox while logging in.
`current_user` will then pick either `session` or the permanent `cookies`.
Please may you elaborate more to the change

File: app/models/user.rb (hunk 1 above)
Comment (OlukaDenis) [reply app/models/user.rb:45]: This works well from my end

**Review by Ispirett — APPROVED** [review]
Status approved :+1: 
Great work @OlukaDenis  :100:  you can move on now!

**Pull request merged**